### ISSUE TITLE
Error: 'Network.setBlockedURLs' wasn't found

### ISSUE THREAD

**Issue by voidmonk** [cyrus-and/chrome-remote-interface#129]
Updated package from 0.18 to 0.22 (latest), and it looks like `Network.addBlockedURL` has been replaced with `Network.setBlockedURLs` in CDP, but on Chrome 58 (latest stable; Windows) its throwing an error:

> UnhandledPromiseRejectionWarning: Unhandled promise rejection
> Error: 'Network.setBlockedURLs' wasn't found

Any clues?!

**Comment by paambaati** [comment]
[`Network.setBlockedURLs`](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-setBlockedURLs) is still marked as experimental. It works for me on Chrome >= 59.

Can you try one of the dev/beta/canary builds?

**Comment by voidmonk** [comment]
Hmm, I'll revert back to 0.18 in that case. Ideally, these npm releases (unless beta or pre-release) shouldn't have non-stable features. It can break working code & builds.

**Comment by cyrus-and** [comment]
@voidmonk the `protocol.json` that comes with this module (ant that's used by default) is just an *approximation* it may be too new or to old with respect to your Chrome version, see the [README][1].

> shouldn't have non-stable or incompatible features

This is just an interface, the non-stable features are in Chrome, where everything that's maked as experimental may change at any time,.

It's impossible for me to ship with a protocol version that works for everyone, for this reason there are workarounds (see the link above). Reverting in this case is never a solution. The recommended solution is to manually invoke those methods that don't match the local `protocol.json`:

```js
client.send('Network.setBlockedURLs', {...});
```

You can even dynamically fetch the proper protocol version using the `remote` option but keep in mind that it [currently][2] involves a couple of HTTP requests.

[1]: https://github.com/cyrus-and/chrome-remote-interface#chrome-debugging-protocol-versions
[2]: https://bugs.chromium.org/p/chromium/issues/detail?id=538300

**Comment by paulirish** [comment]
i'm 70% sure but i think addBlockedURL was in m58 but not m59, where setBlockedURLs took over.
So right now 58 needs a fallback approach.

but its possible to test with setBlockedURLs and see if it fails and then use the old one.

**Comment by voidmonk** [comment]
I understand your views guys, but for an end-user all it boils down to is that the stable release of an interface lib doesn't work with the stable release of its upstream software.

@paulirish `Network.addBlockedURL` works in m58, but the latest CDP `protocol.json` doesn't have it anymore, so it can't be referenced via latest CRI. I still think `Network.addBlockedURL` shouldn't have been removed from `protocol.json` by the Chrome team, until it's removed from Chrome stable.

Maybe it needs some thought on how to mitigate such breaking issues in the future, specially as more people will start to use CRI and Chrome headless.

**Comment by cyrus-and** [comment]
@voidmonk

> but for an end-user all it boils down to is that the stable release of an interface lib doesn't work with the stable release of its upstream software.

This is not accurate, what you're trying to use is marked as experimental, plus it's on the [*tot*][1] branch: *"The latest (tip-of-tree) protocol changes frequently and may break at any time."* And this has nothing to do with the stable release of Chrome or of this library.

I cannot use the `protocol.json` of the stable Chrome as many developers use the beta/unstable version. But that's a choice really, there is not a right answer here, for this reason, as I said, one can bypass the local version of `protocol.json` in many ways.

[1]: https://chromedevtools.github.io/devtools-protocol/tot/

> client.send('Network.setBlockedURLs', {...}); throws the same exception

Yeah, sorry, misread your error, of course the error it's the same. Anyway this time it has nothing to do with `protocol.json`, in fact my answer was about how to use methods and events ignoring the local `protocol.json`.

> but the latest protocol.json has a reference to it

I think you're misunderstanding this point, when you use `client.send()` you're bypassing the local `protocol.json`.

---

**In short**, you're using Chrome 58, you need to use `Network.setBlockedURLs` but it's not supported by your version of Chrome, so you need to use `Network.addBlockedURL`, but the local version of `protocol.json` doesn't have it, so you either invoke `clien.send('Network.addBlockedURL')` or use the `remote` option.
